File: src/spirite/pc/PCUtil.java
package spirite.pc;

import com.jogamp.opengl.GL2;
import spirite.base.graphics.gl.GLEngine;
import spirite.base.graphics.gl.GLImage;
import spirite.base.util.linear.MutableTransform;
import spirite.base.util.linear.Transform;
import spirite.hybrid.HybridHelper;
import spirite.pc.jogl.JOGLCore;
import sun.awt.image.IntegerInterleavedRaster;

import java.awt.geom.AffineTransform;
import java.awt.image.BufferedImage;
import java.awt.image.WritableRaster;
import java.nio.IntBuffer;

public class PCUtil {
	public static BufferedImage deepCopy( BufferedImage toCopy) {
		return new BufferedImage( 
				toCopy.getColorModel(),
				toCopy.copyData(null),
				toCopy.isAlphaPremultiplied(),
				null);
	}
	
	public static BufferedImage glToBI( GLImage img) {
		GLEngine engine = GLEngine.getInstance();
		engine.setTarget(img);
		BufferedImage bi = glSurfaceToImage( HybridHelper.BI_FORMAT, img.getWidth(), img.getHeight());
		engine.setTarget(0);
		return bi;
	}
	
	/** Writes the active GL Surface to a BufferedImage */
	public static BufferedImage glSurfaceToImage(int type, int width, int height) {
		GL2 gl = JOGLCore.getGL2();
		BufferedImage bi = null;
		
		switch( type) {
		case BufferedImage.TYPE_INT_ARGB:
		case BufferedImage.TYPE_INT_ARGB_PRE:
			bi = new BufferedImage(width, height, type);
			
			IntegerInterleavedRaster iir = (IntegerInterleavedRaster)bi.getRaster();
			IntBuffer ib = IntBuffer.wrap(iir.getDataStorage());
			
			gl.glReadPixels( 0, 0, width, height, 
					GL2.GL_BGRA,
					GL2.GL_UNSIGNED_INT_8_8_8_8_REV,
					ib);
		break;
		default: 
			return null;
		}
		
		// Flip Vertically
		if( true) {
			final WritableRaster raster = bi.getRaster();
			Object scanline1 = null;
			Object scanline2 = null;
			for( int i=0; i<bi.getHeight()/2; ++i) {
				scanline1 = raster.getDataElements(0, i, bi.getWidth(), 1, scanline1);
				scanline2 = raster.getDataElements(0, bi.getHeight()-i-1, bi.getWidth(), 1, scanline2);
				raster.setDataElements(0, i, bi.getWidth(), 1, scanline2);
				raster.setDataElements(0, bi.getHeight()-i-1, bi.getWidth(), 1, scanline1);
			}
		}
        
		return bi;
		
	}
	
	/** Converts a MatTrans to an AffineTransform */
	public static AffineTransform toAT( Transform trans) {
		return new AffineTransform(
				trans.getM00(), trans.getM10(), trans.getM01(),
				trans.getM11(), trans.getM02(), trans.getM12());
	}
	/** Converts an AffineTransform to a MatTrans */
	public static MutableTransform toMT(AffineTransform trans) {
		return new MutableTransform(
				(float)trans.getScaleX(), (float)trans.getShearX(), (float)trans.getTranslateX(),
				(float)trans.getShearY(), (float)trans.getScaleY(), (float)trans.getTranslateY());
	}
}

File: src/main/kotlin/spirite/pc/PCUtil.kt
package spirite.pc

import jspirite.rasters.RasterHelper
import spirite.base.graphics.gl.GLEngine
import spirite.base.graphics.gl.GLImage
import spirite.base.util.glu.GLC
import spirite.base.util.linear.MutableTransform
import spirite.base.util.linear.Transform
import spirite.pc.JOGL.JOGL.JOGLInt32Source
import java.awt.geom.AffineTransform
import java.awt.image.BufferedImage
import java.nio.IntBuffer

fun BufferedImage.deepCopy() = BufferedImage(
        this.colorModel,
        this.copyData(null),
        this.isAlphaPremultiplied,
        null)

fun GLImage.toBufferedImage() : BufferedImage {
    val gle = this.engine
    gle.setTarget( this)

    val format = if( this.premultiplied) BufferedImage.TYPE_INT_ARGB_PRE else BufferedImage.TYPE_INT_ARGB

    val bi = gle.surfaceToBufferedImage( format, this.width, this.height)
    //gle.setTarget(null)   // Shouldn't be necessary
    return bi
}

fun GLEngine.surfaceToBufferedImage( type: Int, width: Int, height: Int) : BufferedImage{
    val bi = when( type) {
        BufferedImage.TYPE_INT_ARGB,
        BufferedImage.TYPE_INT_ARGB_PRE -> {
            val bi = BufferedImage(width, height, type)

            val internalStorage = RasterHelper.GetDataStorageFromBi(bi) as IntArray
            val ib = IntBuffer.wrap( internalStorage)

            getGl().readPixels( 0, 0, width, height,
                    GLC.BGRA,
                    GLC.UNSIGNED_INT_8_8_8_8_REV,
                    JOGLInt32Source(ib))

            bi
        }
        else -> BufferedImage(1, 1, type)
    }

    return bi
}

/** Converts a MatTrans to an AffineTransform  */
fun Transform.toAT(): AffineTransform {
    return AffineTransform(
            this.m00, this.m10, this.m01,
            this.m11, this.m02, this.m12)
}

/** Converts an AffineTransform to a MatTrans  */
fun AffineTransform.toMT( ): MutableTransform {
    return MutableTransform(
            this.scaleX.toFloat(), this.shearX.toFloat(), this.translateX.toFloat(),
            this.shearY.toFloat(), this.scaleY.toFloat(), this.translateY.toFloat())
}